packages/application-extension/src/index.tsx

```diff
@@ -430,111 +438,237 @@ const sidebar: JupyterFrontEndPlugin<void> = {
+      widget => widget.id === node.dataset.id
+    );
+
+    if (matches.length < 1) {
+      return shell.currentWidget;
+    }
+
+    return matches[0];
+  };
+
+  // Closes an array of widgets.
+  const closeWidgets = (widgets: Array<Widget>): void => {
+    widgets.forEach(widget => widget.close());
+  };
+
+  // Find the tab area for a widget within a specific dock area.
+  const findTab = (
+    area: DockLayout.AreaConfig,
+    widget: Widget
+  ): DockLayout.ITabAreaConfig | null => {
+    switch (area.type) {
+      case 'split-area':
+        const iterator = iter(area.children);
+        let tab: DockLayout.ITabAreaConfig | null = null;
+        let value: DockLayout.AreaConfig | null = null;
+        do {
+          value = iterator.next();
+          if (value) {
+            tab = findTab(value, widget);
+          }
+        } while (!tab && value);
+        return tab;
+      case 'tab-area':
+        const { id } = widget;
+        return area.widgets.some(widget => widget.id === id) ? area : null;
+      default:
+        return null;
+    }
+  };
+
+  // Find the tab area for a widget within the main dock area.
+  const tabAreaFor = (widget: Widget): DockLayout.ITabAreaConfig | null => {
+    const { mainArea } = shell.saveLayout();
+    if (mainArea.mode !== 'multiple-document') {
+      return null;
+    }
+    let area = mainArea.dock.main;
+    if (!area) {
+      return null;
+    }
+    return findTab(area, widget);
+  };
+
+  // Returns an array of all widgets to the right of a widget in a tab area.
+  const widgetsRightOf = (widget: Widget): Array<Widget> => {
+    const { id } = widget;
+    const tabArea = tabAreaFor(widget);
+    const widgets = tabArea ? tabArea.widgets || [] : [];
+    const index = widgets.findIndex(widget => widget.id === id);
+    if (index < 0) {
+      return [];
+    }
+    return widgets.slice(index + 1);
+  };
+
+  commands.addCommand(CommandIDs.activateNextTab, {
     label: 'Activate Next Tab',
     execute: () => {
-      app.shell.activateNextTab();
+      shell.activateNextTab();
     }
   });
-  palette.addItem({ command, category });
+  palette.addItem({ command: CommandIDs.activateNextTab, category });
 
-  command = CommandIDs.activatePreviousTab;
-  app.commands.addCommand(command, {
+  commands.addCommand(CommandIDs.activatePreviousTab, {
     label: 'Activate Previous Tab',
     execute: () => {
-      app.shell.activatePreviousTab();
+      shell.activatePreviousTab();
+    }
+  });
+  palette.addItem({ command: CommandIDs.activatePreviousTab, category });
+
+  commands.addCommand(CommandIDs.close, {
+    label: () => 'Close Tab',
+    isEnabled: () =>
+      !!shell.currentWidget && !!shell.currentWidget.title.closable,
+    execute: () => {
+      if (shell.currentWidget) {
+        shell.currentWidget.close();
+      }
     }
   });
-  palette.addItem({ command, category });
+  palette.addItem({ command: CommandIDs.close, category });
```

Comment:
I think we should add this command to the context menu as well.

Reply:
Done!